src/commands/preview/dev_server.rs

```diff
@@ -0,0 +1,190 @@
+
+        let mut address_iter = try_address.to_socket_addrs()?;
+
+        let listening_address = address_iter
+            .next()
+            .ok_or_else(|| format_err!("Could not parse address {}", try_address))?;
+
+        let host = host.unwrap_or("https://example.com").to_string();
+
+        let parsed_url = match Url::parse(&host) {
+            Ok(host) => Ok(host),
+            Err(_) => Url::parse(&format!("https://{}", host)),
+        }?;
+
+        let scheme = parsed_url.scheme();
+        if scheme != "http" && scheme != "https" {
+            failure::bail!("Your host scheme must be either http or https")
+        }
+        let is_https = scheme == "https";
+
+        let host = parsed_url.host_str().ok_or(format_err!("Invalid host, accepted formats are example.com, http://example.com, or https://example.com"))?.to_string();
+
+        Ok(ServerConfig {
+            listening_address,
+            host,
+            is_https,
+        })
+    }
+
+    fn listening_address_as_string(&self) -> String {
+        self.listening_address
+            .to_string()
+            .replace("[::1]", "localhost")
+    }
+}
+
+pub async fn dev_server(
+    target: Target,
+    user: Option<GlobalUser>,
+    host: Option<&str>,
+    port: Option<&str>,
+    ip: Option<&str>,
+) -> Result<(), failure::Error> {
+    commands::build(&target)?;
+    let server_config = ServerConfig::new(host, ip, port)?;
+    let https = HttpsConnector::new().expect("TLS initialization failed");
+    let client = Client::builder().build::<_, Body>(https);
+
+    let preview_id = get_preview_id(target, user, &server_config)?;
+    let listening_address = server_config.listening_address.clone();
+    let listening_address_string = server_config.listening_address_as_string();
+
+    let make_service = make_service_fn(move |_| {
+        let client = client.clone();
+        let preview_id = preview_id.to_owned();
+        let server_config = server_config.clone();
+        async move {
+            Ok::<_, failure::Error>(service_fn(move |req| {
+                preview_request(
+                    req,
+                    client.to_owned(),
+                    preview_id.to_owned(),
+                    server_config.clone(),
+                )
+            }))
+        }
+    });
+
+    let server = Server::bind(&listening_address).serve(make_service);
+    println!(
+        "{} Listening on http://{}",
+        emoji::EAR,
+        listening_address_string
+    );
+    if let Err(e) = server.await {
+        eprintln!("server error: {}", e);
+    }
+    Ok(())
+}
+
+fn get_preview_url(path_string: &str) -> Result<Uri, http::uri::InvalidUri> {
+    format!("https://{}{}", PREVIEW_HOST, path_string).parse()
+}
+
+fn get_path_as_str(uri: Uri) -> String {
+    uri.path_and_query()
+        .map(|x| x.as_str())
+        .unwrap_or("")
+        .to_string()
+}
+
+fn preview_request(
+    req: Request<Body>,
+    client: Client<HttpsConnector<HttpConnector>>,
```

Comment:
not that i am opposed, but this appears to introduce a new client

Reply:
this reuses the same client we created in the other method - it's a parameter